docs/content/sips/002-app-config.md

```diff
@@ -0,0 +1,141 @@
+title = "SIP xxx - Application Configuration"
+template = "main"
+date = "2022-03-22T14:53:30Z"
+
+---
+
+Summary: A configuration system for Spin applications.
+
+Owner: lann.martin@fermyon.com
+
+Created: March 22, 2022
+
+Updated: April 4, 2022
+
+## Background
+
+It is common for applications to require configuration at runtime that isn't known at build time or is too sensitive to be stored with build artifacts:
+
+- Logging configuration
+- Per-channel (production, staging, etc) service dependency URLs
+- Database secrets
+
+## Proposal
+
+### Configuration schema is defined by components and applications
+
+Configuration within a "parent" (component or application) consists of a number of configuration "slots":
+- Slots are identified by a string "key"; in order to allow unambiguous conversion to environment variables or file paths, keys are constrained:
+  - Keys must start with a letter (required for env vars)
+  - Keys consist of only lowercase ascii alphanum and `_` (`[a-z0-9_]`)
+    - Only one `_` at a time and not at the end (to allow delimiting in env vars with `__`)
+- Slot keys must be unique within their parent, but to allow independent development of components different parents may have identical keys
+- A slot must _either_ be marked as "required" _or_ must be given a default value
+- A slot may be marked as "secret", in which case any associated value should be handled with care (e.g. not logged)
+
+```toml
+[config]
+required_key = { required = true }
+optional_key = { default = "default_value" }
+secret_key = { required = true, secret = true }
+```
+
+Default values can use template strings to reference other slots.
+```toml
+[config]
+key1 = { required = true }
+key2 = { default = "prefix-{{ key1 }}-suffix" }
+```
+
+### Components and applications set configuration values of their dependencies
+
+In dependency configuration, templates strings can reference top-level config keys (those in `[config]`), "sibling" keys within the same dependency, and "ancestor" dependant configs.
+
+- Top-level references use just the key name: `{{ top_level_key }}`
+- "Sibling" references use a single `.` prefix: `{{ .sibling_key }}`
+- "Ancestor" references use multiple `.`s: `{{ ..parent_dep_key }}`, `{{ ...grandparent_key }}`
+- Circular / infinitely recursive references are not permitted
+
+`spin.toml`:
+```toml
+[config]
+app_root = { default = "/app" }
+log_file = "{{ app_root }}/log.txt"    # -> "/app/log.txt"
```

Comment:
Is this a fixed value, or a sneaky instance of the old syntax for defaults?

Reply:
dagnabbit